File: javascript/util/node/web_socket/protocol8_parser.js
Faye.WebSocket.Protocol8Parser = Faye.Class({
  <% %w[FIN MASK RSV1 RSV2 RSV3 OPCODE LENGTH].each do |const| %>
    <%= const %>: <%= Faye::WebSocket::Protocol8Parser.const_get(const) %>,
  <% end %>
  
  GUID: '<%= Faye::WebSocket::Protocol8Parser::GUID %>',
  
  OPCODES: <%= JSON.dump Faye::WebSocket::Protocol8Parser::OPCODES %>,
  FRAGMENTED_OPCODES: <%= JSON.dump Faye::WebSocket::Protocol8Parser::FRAGMENTED_OPCODES %>,
  OPENING_OPCODES: <%= JSON.dump Faye::WebSocket::Protocol8Parser::OPENING_OPCODES %>,
  
  ERRORS: <%= JSON.dump Faye::WebSocket::Protocol8Parser::ERRORS %>,
  
  ERROR_CODES: <%= JSON.dump Faye::WebSocket::Protocol8Parser::ERROR_CODES %>,
  
  UTF8_MATCH: <%= Faye::UTF8_MATCH.inspect %>,
  
  version: 'protocol-8',
  
  Handshake: Faye.Class({
    initialize: function(uri, stream) {
      this._uri    = uri;
      this._stream = stream;
      
      var buffer = new Buffer(16), i = 16;
      while (i--) buffer[i] = Math.floor(Math.random() * 254);
      this._key = buffer.toString('base64');
      
      var SHA1 = crypto.createHash('sha1');
      SHA1.update(this._key + Faye.WebSocket.Protocol8Parser.prototype.GUID);
      this._accept = SHA1.digest('base64');
      
      var HTTPParser = process.binding('http_parser').HTTPParser,
          parser     = new HTTPParser('response'),
          current    = null,
          self       = this;
      
      this._complete = false;
      this._headers  = {};
      this._parser   = parser;
      
      parser.onHeaderField = function(b, start, length) {
        current = b.toString('utf8', start, start + length);
      };
      parser.onHeaderValue = function(b, start, length) {
        self._headers[current] = b.toString('utf8', start, start + length);
      };
      parser.onHeadersComplete = function(settings) {
        self._status = settings.statusCode;
      };
      parser.onMessageComplete = function() {
        self._complete = true;
      };
    },
    
    requestData: function() {
      var stream = this._stream;
      try {
        stream.write('GET ' + this._uri.pathname + ' HTTP/1.1\r\n');
        stream.write('Host: ' + this._uri.hostname + '\r\n');
        stream.write('Upgrade: websocket\r\n');
        stream.write('Connection: Upgrade\r\n');
        stream.write('Sec-WebSocket-Key: ' + this._key + '\r\n');
        stream.write('Sec-WebSocket-Version: 8\r\n');
        stream.write('\r\n');
      } catch (e) {}
    },
    
    parse: function(data) {
      this._parser.execute(data, 0, data.length);
    },
    
    isComplete: function() {
      return this._complete;
    },
    
    isValid: function() {
      if (this._status !== 101) return false;
      
      var upgrade    = this._headers.Upgrade,
          connection = this._headers.Connection;
      
      return upgrade && /^websocket$/i.test(upgrade) &&
             connection && connection.split(/\s*,\s*/).indexOf('Upgrade') >= 0 &&
             this._headers['Sec-WebSocket-Accept'] === this._accept;
    }
  }),
  
  initialize: function(webSocket, stream) {
    this._reset();
    this._socket = webSocket;
    this._stream = stream;
    this._stage  = 0;
  },
  
  handshakeResponse: function() {
    var secKey = this._socket.request.headers['sec-websocket-key'];
    if (!secKey) return;
    
    var SHA1 = crypto.createHash('sha1');
    SHA1.update(secKey + this.GUID);
    var accept = SHA1.digest('base64');
    
    var stream = this._stream;
    try {
      stream.write('HTTP/1.1 101 Switching Protocols\r\n');
      stream.write('Upgrade: websocket\r\n');
      stream.write('Connection: Upgrade\r\n');
      stream.write('Sec-WebSocket-Accept: ' + accept + '\r\n\r\n');
    } catch (e) {
      // socket closed while writing
      // no handshake sent; client will stop using WebSocket
    }
  },
  
  createHandshake: function() {
    return new this.Handshake(this._socket.uri, this._stream);
  },
  
  parse: function(data) {
    for (var i = 0, n = data.length; i < n; i++) {
      switch (this._stage) {
        case 0: this._parseOpcode(data[i]);         break;
        case 1: this._parseLength(data[i]);         break;
        case 2: this._parseExtendedLength(data[i]); break;
        case 3: this._parseMask(data[i]);           break;
        case 4: this._parsePayload(data[i]);        break;
      }
      if (this._stage === 4 && this._length === 0)
        this._emitFrame();
    }
  },
  
  frame: function(data, type, errorType) {
    if (this._closed) return;
    
    var opcode = this.OPCODES[type || (typeof data === 'string' ? 'text' : 'binary')],
        buffer = new Buffer(data),
        error  = this.ERRORS[errorType],
        insert = error ? 2 : 0,
        length = buffer.length + insert,
        stream = this._stream,
        frame, factor;
    
    if (length <= 125) {
      frame = new Buffer(2 + insert);
      frame[1] = length;
    } else if (length >= 126 && length <= 65535) {
      frame = new Buffer(4 + insert);
      frame[1] = 126;
      frame[2] = Math.floor(length / 256);
      frame[3] = length & 255;
    } else {
      frame = new Buffer(10 + insert);
      frame[1] = 127;
      for (var i = 0; i < 8; i++) {
        factor = Math.pow(2, 8 * (8 - 1 - i));
        frame[2+i] = Math.floor(length / factor) & 255;
      }
    }
    frame[0] = this.FIN | opcode;
    
    if (error) {
      frame[frame.length - 2] = Math.floor(error / 256);
      frame[frame.length - 1] = error & 255;
    }
    
    try {
      stream.write(frame, 'binary');
      if (buffer.length > 0) stream.write(buffer, 'utf8');
      return true;
    } catch (e) {
      return false;
    }
  },
  
  close: function(errorType, callback, context) {
    if (this._closed) return;
    if (callback) this._closingCallback = [callback, context];
    this.frame('', 'close', errorType || 'normal_closure');
    this._closed = true;
  },
  
  buffer: function(fragment) {
    for (var i = 0, n = fragment.length; i < n; i++)
      this._buffer.push(fragment[i]);
  },
  
  _parseOpcode: function(data) {
    var rsvs = [this.RSV1, this.RSV2, this.RSV3].filter(function(rsv) {
      return (data & rsv) === rsv;
    }, this);
    
    if (rsvs.length > 0) return this._socket.close('protocol_error');
    
    this._final   = (data & this.FIN) === this.FIN;
    this._opcode  = (data & this.OPCODE);
    this._mask    = [];
    this._payload = [];
    
    var valid = false;
    
    for (var key in this.OPCODES) {
      if (this.OPCODES[key] === this._opcode)
        valid = true;
    }
    if (!valid) return this._socket.close('protocol_error');
    
    if (Faye.indexOf(this.FRAGMENTED_OPCODES, this._opcode) < 0 && !this._final)
      return this._socket.close('protocol_error');
    
    if (this._mode && Faye.indexOf(this.OPENING_OPCODES, this._opcode) >= 0)
      return this._socket.close('protocol_error');
    
    this._stage = 1;
  },
  
  _parseLength: function(data) {
    this._masked = (data & this.MASK) === this.MASK;
    this._length = (data & this.LENGTH);
    
    if (this._length >= 0 && this._length <= 125) {
      this._stage = this._masked ? 3 : 4;
    } else {
      this._lengthBuffer = [];
      this._lengthSize   = (this._length === 126 ? 2 : 8);
      this._stage        = 2;
    }
  },
  
  _parseExtendedLength: function(data) {
    this._lengthBuffer.push(data);
    if (this._lengthBuffer.length < this._lengthSize) return;
    this._length = this._getInteger(this._lengthBuffer);
    this._stage  = this._masked ? 3 : 4;
  },
  
  _parseMask: function(data) {
    this._mask.push(data);
    if (this._mask.length < 4) return;
    this._stage = 4;
  },
  
  _parsePayload: function(data) {
    this._payload.push(data);
    if (this._payload.length < this._length) return;
    this._emitFrame();
  },
  
  _emitFrame: function() {
    var payload = this._unmask(this._payload, this._mask),
        opcode  = this._opcode;
    
    if (opcode === this.OPCODES.continuation) {
      if (!this._mode) return this._socket.close('protocol_error');
      this.buffer(payload);
      if (this._final) {
        var message = new Buffer(this._buffer);
        if (this._mode === 'text') message = this._encode(message);
        this._reset();
        if (message !== null) this._socket.receive(message);
        else this._socket.close('encoding_error');
      }
    }
    else if (opcode === this.OPCODES.text) {
      if (this._final) {
        var message = this._encode(payload);
        if (message !== null) this._socket.receive(message);
        else this._socket.close('encoding_error');
      } else {
        this._mode = 'text';
        this.buffer(payload);
      }
    }
    else if (opcode === this.OPCODES.binary) {
      if (this._final) {
        this._socket.receive(payload);
      } else {
        this._mode = 'binary';
        this.buffer(payload);
      }
    }
    else if (opcode === this.OPCODES.close) {
      var errorCode = (payload.length >= 2) ? 256 * payload[0] + payload[1] : 0,
          
          errorType = (payload.length === 0) ||
                      (errorCode >= 3000 && errorCode < 5000) ||
                      Faye.indexOf(this.ERROR_CODES, errorCode) >= 0 ?
                      'normal_closure' :
                      'protocol_error';
      
      if (payload.length > 125 || (payload.length > 2 && !this._encode(payload.slice(2))))
        errorType = 'protocol_error';
      
      this._socket.close(errorType);
      if (this._closingCallback)
        this._closingCallback[0].call(this._closingCallback[1]);
    }
    else if (opcode === this.OPCODES.ping) {
      if (payload.length > 125) return this._socket.close('protocol_error');
      this._socket.send(payload, 'pong');
    }
    this._stage = 0;
  },
  
  _reset: function() {
    this._mode   = null;
    this._buffer = [];
  },
  
  _encode: function(buffer) {
    try {
      var string = buffer.toString('binary', 0, buffer.length);
      if (!this.UTF8_MATCH.test(string)) return null;
    } catch (e) {}
    return buffer.toString('utf8', 0, buffer.length);
  },
  
  _getInteger: function(bytes) {
    var number = 0;
    for (var i = 0, n = bytes.length; i < n; i++)
      number += bytes[i] << (8 * (n - 1 - i));
    return number;
  },
  
  _unmask: function(payload, mask) {
    var unmasked = new Buffer(payload.length), b;
    for (var i = 0, n = payload.length; i < n; i++) {
      b = payload[i];
      if (mask.length > 0) b = b ^ mask[i % 4];
      unmasked[i] = b;
    }
    return unmasked;
  }
});

File: lib/faye/util/web_socket/protocol8_parser.rb
require 'base64'
require 'digest/sha1'
require 'net/http'

module Faye
  class WebSocket
    
    class Protocol8Parser
      GUID = '258EAFA5-E914-47DA-95CA-C5AB0DC85B11'
      
      FIN = MASK = 0b10000000
      RSV1       = 0b01000000
      RSV2       = 0b00100000
      RSV3       = 0b00010000
      OPCODE     = 0b00001111
      LENGTH     = 0b01111111
      
      OPCODES = {
        :continuation => 0,
        :text         => 1,
        :binary       => 2,
        :close        => 8,
        :ping         => 9,
        :pong         => 10
      }
      
      FRAGMENTED_OPCODES = OPCODES.values_at(:continuation, :text, :binary)
      OPENING_OPCODES = OPCODES.values_at(:text, :binary)
      
      ERRORS = {
        :normal_closure   => 1000,
        :going_away       => 1001,
        :protocol_error   => 1002,
        :unacceptable     => 1003,
        :encoding_error   => 1007,
        :policy_violation => 1008,
        :too_large        => 1009,
        :extension_error  => 1010
      }
      
      ERROR_CODES = ERRORS.values
      
      class Handshake
        def initialize(uri)
          @uri    = uri
          @key    = Base64.encode64((1..16).map { rand(255).chr } * '').strip
          @accept = Base64.encode64(Digest::SHA1.digest(@key + GUID)).strip
          @buffer = []
        end
        
        def request_data
          hostname = @uri.host + (@uri.port ? ":#{@uri.port}" : '')
          
          handshake  = "GET #{@uri.path}#{@uri.query ? '?' : ''}#{@uri.query} HTTP/1.1\r\n"
          handshake << "Host: #{hostname}\r\n"
          handshake << "Upgrade: websocket\r\n"
          handshake << "Connection: Upgrade\r\n"
          handshake << "Sec-WebSocket-Key: #{@key}\r\n"
          handshake << "Sec-WebSocket-Version: 8\r\n"
          handshake << "\r\n"
          
          handshake
        end
        
        def parse(data)
          message  = []
          complete = false
          data.each_byte do |byte|
            if complete
              message << byte
            else
              @buffer << byte
              complete ||= complete?
            end
          end
          message
        end
        
        def complete?
          @buffer[-4..-1] == [0x0D, 0x0A, 0x0D, 0x0A]
        end
        
        def valid?
          data = Faye.encode(@buffer)
          response = Net::HTTPResponse.read_new(Net::BufferedIO.new(StringIO.new(data)))
          return false unless response.code.to_i == 101
          
          upgrade, connection = response['Upgrade'], response['Connection']
          
          upgrade and upgrade =~ /^websocket$/i and
          connection and connection.split(/\s*,\s*/).include?('Upgrade') and
          response['Sec-WebSocket-Accept'] == @accept
        end
      end
      
      def initialize(web_socket, options = {})
        reset
        @socket  = web_socket
        @stage   = 0
        @masking = options[:masking]
      end
      
      def version
        'protocol-8'
      end
      
      def handshake_response
        sec_key = @socket.env['HTTP_SEC_WEBSOCKET_KEY']
        return '' unless String === sec_key
        
        accept = Base64.encode64(Digest::SHA1.digest(sec_key + GUID)).strip
        
        upgrade =  "HTTP/1.1 101 Switching Protocols\r\n"
        upgrade << "Upgrade: websocket\r\n"
        upgrade << "Connection: Upgrade\r\n"
        upgrade << "Sec-WebSocket-Accept: #{accept}\r\n"
        upgrade << "\r\n"
        upgrade
      end
      
      def create_handshake
        Handshake.new(@socket.uri)
      end
      
      def parse(data)
        data.each_byte do |byte|
          case @stage
          when 0 then parse_opcode(byte)
          when 1 then parse_length(byte)
          when 2 then parse_extended_length(byte)
          when 3 then parse_mask(byte)
          when 4 then parse_payload(byte)
          end
          emit_frame if @stage == 4 and @length == 0
        end
      end
      
      def frame(data, type = nil, code = nil)
        return nil if @closed
        
        type ||= (String === data ? :text : :binary)
        data   = data.bytes.to_a if data.respond_to?(:bytes)
        
        if code
          data = [code].pack('n').bytes.to_a + data
        end
        
        frame  = (FIN | OPCODES[type]).chr
        length = data.size
        masked = @masking ? MASK : 0
        
        case length
          when 0..125 then
            frame << (masked | length).chr
          when 126..65535 then
            frame << (masked | 126).chr
            frame << [length].pack('n')
          else
            frame << (masked | 127).chr
            frame << [length >> 32, length & 0xFFFFFFFF].pack('NN')
        end
        
        if @masking
          mask = (1..4).map { rand 256 }
          data.each_with_index do |byte, i|
            data[i] = byte ^ mask[i % 4]
          end
          frame << mask.pack('C*')
        end
        
        Faye.encode(frame) + Faye.encode(data)
      end
      
      def close(code = nil, reason = nil, &callback)
        return if @closed
        @closing_callback ||= callback
        @socket.send(reason || '', :close, code || ERRORS[:normal_closure])
        @closed = true
      end
      
    private
      
      def parse_opcode(data)
        if [RSV1, RSV2, RSV3].any? { |rsv| (data & rsv) == rsv }
          return @socket.close(ERRORS[:protocol_error], nil, false)
        end
        
        @final   = (data & FIN) == FIN
        @opcode  = (data & OPCODE)
        @mask    = []
        @payload = []
        
        unless OPCODES.values.include?(@opcode)
          return @socket.close(ERRORS[:protocol_error], nil, false)
        end
        
        unless FRAGMENTED_OPCODES.include?(@opcode) or @final
          return @socket.close(ERRORS[:protocol_error], nil, false)
        end
        
        if @mode and OPENING_OPCODES.include?(@opcode)
          return @socket.close(ERRORS[:protocol_error], nil, false)
        end
        
        @stage = 1
      end
      
      def parse_length(data)
        @masked = (data & MASK) == MASK
        @length = (data & LENGTH)
        
        if @length <= 125
          @stage = @masked ? 3 : 4
        else
          @length_buffer = []
          @length_size   = (@length == 126) ? 2 : 8
          @stage         = 2
        end
      end
      
      def parse_extended_length(data)
        @length_buffer << data
        return unless @length_buffer.size == @length_size
        @length = integer(@length_buffer)
        @stage  = @masked ? 3 : 4
      end
      
      def parse_mask(data)
        @mask << data
        return if @mask.size < 4
        @stage = 4
      end
      
      def parse_payload(data)
        @payload << data
        return if @payload.size < @length
        emit_frame
      end
      
      def emit_frame
        payload = unmask(@payload, @mask)
        
        case @opcode
          when OPCODES[:continuation] then
            return @socket.close(ERRORS[:protocol_error], nil, false) unless @mode
            @buffer += payload
            if @final
              message = @buffer
              message = Faye.encode(message, true) if @mode == :text
              reset
              if message
                @socket.receive(message)
              else
                @socket.close(ERRORS[:encoding_error], nil, false)
              end
            end

          when OPCODES[:text] then
            if @final
              message = Faye.encode(payload, true)
              if message
                @socket.receive(message)
              else
                @socket.close(ERRORS[:encoding_error], nil, false)
              end
            else
              @mode = :text
              @buffer += payload
            end

          when OPCODES[:binary] then
            if @final
              @socket.receive(payload)
            else
              @mode = :binary
              @buffer += payload
            end

          when OPCODES[:close] then
            code = (payload.size >= 2) ? 256 * payload[0] + payload[1] : nil
            
            unless (payload.size == 0) or
                   (code && code >= 3000 && code < 5000) or
                   ERROR_CODES.include?(code)
              code = ERRORS[:protocol_error]
            end
            
            if payload.size > 125 or not Faye.valid_utf8?(payload[2..-1] || [])
              code = ERRORS[:protocol_error]
            end
            
            reason = (payload.size > 2) ? Faye.encode(payload[2..-1], true) : nil
            @socket.close(code, reason, false)
            @closing_callback.call if @closing_callback

          when OPCODES[:ping] then
            return @socket.close(ERRORS[:protocol_error], nil, false) if payload.size > 125
            @socket.send(payload, :pong)
        end
        @stage = 0
      end

      def reset
        @buffer = []
        @mode   = nil
      end
      
      def integer(bytes)
        number = 0
        bytes.each_with_index do |data, i|
          number += data << (8 * (bytes.size - 1 - i))
        end
        number
      end
      
      def unmask(payload, mask)
        unmasked = []
        payload.each_with_index do |byte, i|
          byte = byte ^ mask[i % 4] if mask.size > 0
          unmasked << byte
        end
        unmasked
      end
    end
    
  end
end

